### PR TITLE
Diagnostic caching prototype

### PR DESCRIPTION
**Pull request by genlu**
@mavasani @CyrusNajmabadi This is behind an experiment.

### CODE REVIEW

**Comment by mavasani** [comment]
@genlu Would you be able to take a small video or set of screenshots for the new experience with the cached diagnostics? I'd like to see the experience when say cached diagnostics are replaced by live diagnostics, or if user starts typing after seeing cached diagnostics, but before live diagnostics/solution crawler has kicked off, etc.

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                return new NoOpDiagnosticCacheService();
+            }
+
+            return new VisualStudioDiagnosticCacheService(vsWorkspace, _diagnosticService, _registrationService, _listenerProvider);
+        }
+
+        private class NoOpDiagnosticCacheService : IDiagnosticCacheService
+        {
+            public Task<bool> TryLoadCachedDiagnosticsAsync(Document document, CancellationToken cancellation)
+                => Task.FromResult(false);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs:56]:
```suggestion
                => SpecializedTasks.False;
```

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        public IWorkspaceService? CreateService(HostWorkspaceServices workspaceServices)
+        {
+            if (workspaceServices.Workspace is not VisualStudioWorkspace vsWorkspace)
+            {
+                return new NoOpDiagnosticCacheService();
+            }
+
+            var experimentationService = workspaceServices.GetRequiredService<IExperimentationService>();
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.DiagnosticCache))
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs:45]: 📝 Experiments need to be backed by both a registry option and an experiment name, otherwise it is impossible to enable the feature even for testing on systems with telemetry disabled. See `OOPServerGC` for examples.

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                    message: string.Format(CachedMessageFormat, d.Message),
+                    enuMessageForBingSearch: d.ENUMessageForBingSearch,
+                    severity: d.Severity, defaultSeverity: d.DefaultSeverity, isEnabledByDefault: d.IsEnabledByDefault, warningLevel: d.WarningLevel,
+                    customTags: d.CustomTags, properties: d.Properties, projectId: d.ProjectId, location: d.DataLocation, additionalLocations: d.AdditionalLocations,
+                    language: d.Language, title: d.Title,
+                    description: d.Description,
+                    helpLink: d.HelpLink, isSuppressed: d.IsSuppressed);
+        }
+
+        private static async Task<Checksum> GetChecksumAsync(Document document, CancellationToken cancellationToken)
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs:113]:
```suggestion
        private static async ValueTask<Checksum> GetChecksumAsync(Document document, CancellationToken cancellationToken)
```

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs [hunk 4]
```diff
@@ -141,6 +136,14 @@ private async Task ProduceTagsAsync(TaggerContext<TTag> context, DocumentSnapsho
             var cancellationToken = context.CancellationToken;
             var workspace = document.Project.Solution.Workspace;
 
+            var diagnosticCacheService = workspace.Services.GetService<IDiagnosticCacheService>();
+            if (diagnosticCacheService != null)
+            {
+                // During solution load. if there's any cached diagnostics available,
+                // this service will (asynchronously) push updates to IDiagnosticService.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs:143]: having hte tagger push to the DiagService seems like a *very* *very* strange data flow arrangement.  Tagging @mavasani .  Iwould not have expected this.  I would prefer we treat the IDiagService as the source of truth, with all our features being driven by the results it produces.

Comment (genlu) [reply src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs:143]: We only need something that's aware of opened document as trigger, it doesn't need to be tagger itself. It can't be solution crawler though, since we are operating before it started. Maybe raw workspace events?

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs (hunk 2 above)
Comment (genlu) [reply src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs]: Added here: https://github.com/dotnet/roslyn/blob/8b11700351e478dfc630e141340d9d23f6d74b8a/src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheServiceFactory.cs#L41

How do I enable it? This command doesn't seem to work
```
vsregedit set local HKCU "Roslyn\Internal\OnOff\Features" DiagnosticCache dword 1
```

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService_ErrorList.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                {
+                    var eventArg = DiagnosticsUpdatedArgs.DiagnosticsCreated(
+                            new CachedDiagnosticsUpdateArgsId(document.Id), document.Project.Solution.Workspace, document.Project.Solution, document.Project.Id, document.Id, diagnostics);
+                    _requestedDocuments.Add(document.Id, eventArg);
+                    _ = _errorListUpdateQueue.ScheduleTask(DiagnosticsUpdatedEventName, () =>
+                    {
+                        if (_diagnosticService is DiagnosticService s)
+                        {
+#pragma warning disable RS0043 // Do not call 'GetTestAccessor()'
+                            s.GetTestAccessor().EventListenerTracker.EnsureEventListener(_workspace, s);
```
Comment (genlu) [inline src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService_ErrorList.cs:60]: @mavasani ErrorList lazily listens to diagnostic service, so it's only started when there's any update from the service. What's the best way to start errorlist before we start pushing cached diagnostics to it?

**Comment by genlu** [comment]
Ping @mavasani for a review

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.DiagnosticCache
+{
+    internal sealed partial class VisualStudioDiagnosticCacheService : IDiagnosticCacheService
+    {
+        // TODO: loc
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs:23]: File issue?

File: src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs [hunk 7]
```diff
@@ -148,24 +154,23 @@ private async Task ProduceTagsAsync(TaggerContext<TTag> context, DocumentSnapsho
...
             buffer?.Properties.TryGetProperty(PredefinedPreviewTaggerKeys.SuppressDiagnosticsSpansKey, out suppressedDiagnosticsSpans);
 
+            var diagnosticCacheService = workspace.Services.GetService<IDiagnosticCacheService>();
+            if (diagnosticCacheService != null)
+            {
+                // During solution load. if there's any cached diagnostics available,
+                // this service will push updates via CachedDiagnosticsUpdated event.
+                await diagnosticCacheService.LoadCachedDiagnosticsAsync(document, cancellationToken).ConfigureAwait(false);
+
+                if (diagnosticCacheService.TryGetLoadedCachedDiagnostics(document.Id, out var cachedId, out var cachedDiagnostics))
```
Comment (mavasani) [inline src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs:164]: Seems little weird to invoke Load method and then a TryGetLoaded method. Maybe better to use an additional parameter `bool loadCachedDiagnostics` for `TryGetLoadedCachedDiagnostics` method?

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService_ErrorList.cs (hunk 5 above)
Comment (mavasani) [reply src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService_ErrorList.cs:63]: @genlu Live diagnostics table source is currently hooked up to be in stable state when solution crawler is not running or has not yet started. You can see the implementation here: https://github.com/dotnet/roslyn/blob/a4bdff76ec342a445d6c59025df96f55688cb8d5/src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseDiagnosticListTable.LiveTableDataSource.cs#L83-L92

IsSolutionCrawlerRunning is implemented here:
https://github.com/dotnet/roslyn/blob/a4bdff76ec342a445d6c59025df96f55688cb8d5/src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractRoslynTableDataSource.cs#L33-L83

Basically, if neither the build nor solution crawler are running, the live diagnostic source will always mark error list stable, which prevents the underlying error list to try to fetch updated diagnostics. You need to change the above logic to also account for cache diagnostics availability.

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                    language: d.Language, title: d.Title,
+                    description: d.Description,
+                    helpLink: d.HelpLink, isSuppressed: d.IsSuppressed);
+        }
+
+        private static async ValueTask<Checksum> GetChecksumAsync(Document document, CancellationToken cancellationToken)
+        {
+            // We only checksum off of the contents of the file.  During load, we can't really compute any other
+            // information since we don't necessarily know about other files, metadata, or dependencies.  So during
+            // load, we allow for the previous diagnostics to be used as long as the file contents match.
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs:132]: This part feels a bit scary - if there were some errors in the prior session due to things like missing references or packages, and the user closed VS, restored and built everything successfully on command line and then re-opened VS, they would still see stale errors from previous session until we replace cached diagnostics with true live diagnostics.

Comment (genlu) [reply src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs:132]: Yeah, this is a very fair concern, any project level change might be problematic in current design. Additionally, I think we could end up caching "wrong" diagnostics in IDE caused by incorrect project restore, which user fixed in commandline w/o changes to projects. I don't have a good solution to this at the moment.

**Comment by CyrusNajmabadi** [comment]
@vatsalyaagrawal What *problem* are we actually trying to solve here?  I imagine problems might be things like:

1. users want to be able to open VS immediately, navigate to a bug quickly, and then fix it.

First, i'm not sure if this 'feature' would even help with the first half.  Since we don't show diagnostics for closed files, you likely won't see the errors, so you won't be able to navigate to it quickly.  Second, we won't be able to help you fix the issue.  If you bring up the lightbulb, VS will just hang until the point that thigns are loaded and real diagnostics are computed.  And, if you edit the code, the error will immediately disappear, but not necessarily be fixed.

The latter bit is very worrying to me.  Imagine i get an error saying "identifier foo cannot be found"  i then change "foo" to "bar" and the error goes away.  However, "bar" is still just as invalid.  Except now the IDE is not telling me that.  From the user perspective we've gone from:

1. we're showing you errors, indicating that things are working properly to
2. we're not showing you errors, but your code may still be in error, indicating that the product is broken

I fundamentally just do not get how this feature provides value.  And it seems to actively come with substantive problems that negate any potential implied value completely.  If VS shows me errors, and then just stops showing me errors even when they're there, i will rightly just assume the product is broken.  I do not think we should add features that cause this sort of user experience confusion.

File: src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService.cs:132]: @vatsalyaagrawal ^ and https://github.com/dotnet/roslyn/pull/50168#issuecomment-784442326
